LTI Consumer xBlock changes to improve the UX in the ASU Gl…

chrisndodge opened the pull request:
…obal Freshman Academy courses

This is the implementation of the previously discussed changes to the LTI Consumer xBlock to support the ASU/GFA courses. There will be a corresponding PR against edx-platform (and I'll add a link to it). 
## **Supporting information**

**Design Document** (reviewer may have to request permission to view it): https://docs.google.com/document/d/1lCqvKL08WtG022Kedl-f3hCnitRq5uXEVwMfUc1fTzE/edit

**sandbox**: 
LMS: http://asu.chrisndodge.com
Studio: http://studio-asu.chrisndodge.com

HTTP Auth/ Application Credentials will be sent separately (since there are LTI secrets in this environment). Please contact author via email.

**corresponding PR to edx-platform**: https://github.com/edx/edx-platform/pull/12997
## **Overview of Changes/Screenshots**

Additional boolean settings fields in the LTI Consumer xBlock:

![screen shot 2016-07-14 at 9 52 51 am](https://cloud.githubusercontent.com/assets/2362917/16842558/e16915e0-49ab-11e6-84dd-d83b5f5ac6d1.png)

Additional HTML settings field in the LTI Consumer xBlock:

![screen shot 2016-07-14 at 9 52 58 am](https://cloud.githubusercontent.com/assets/2362917/16842570/ebf41f0a-49ab-11e6-8120-ccdc42eb6915.png)

Showing HTML settings field as raw HTML:

![screen shot 2016-07-14 at 9 53 04 am](https://cloud.githubusercontent.com/assets/2362917/16842738/7ff6a970-49ac-11e6-96f0-0b0c4c59dd53.png)

Net result that user sees in the LMS, with headers disabled and a custom HTML "launch pane":

![screen shot 2016-07-14 at 9 57 26 am](https://cloud.githubusercontent.com/assets/2362917/16842757/8bd2b158-49ac-11e6-919d-619202224ceb.png)

Launch of the LTI modal window against a "test LTI provider", showing the two new fields added to the LTI launch payload (custom_student_course_mode and custom_student_verification_status):

![screen shot 2016-07-14 at 10 20 49 am](https://cloud.githubusercontent.com/assets/2362917/16842798/b8422660-49ac-11e6-8505-710c7d5c7edf.png)

FYI, @douglashall
 Adding @nwilken, @aricken, and @jdenison72 as observers

robrap commented:
@chrisndodge: I'd first like to pin down the requirements before getting too far into the code.  Here are some questions:

1. In terms of the new LTI launch parameters, is there a reason why they need to be optional?  Did you have a performance concern or other concern?  I'm not sure how I feel about the grouped setting config, and the simplest might be to drop it altogether.

2. It seems like legal is coming down on us providing the visual EXTERNAL RESOURCE indicator.  Depending on the final outcome, we'll need to decide on an implementation that works.  If it is truly a requirement, it should be outside of your "Launch Pane HTML" setting, but what settings will allow for a simple configuration of this without the rest of the Display Header (i.e. a minimal header), especially if ASU needs it as a footer and cares about left/right justification.

3. Could you provide steps/requirements for filling in the "Launch Pane HTML" such that it will work?  It would be nice for there to be some sort of validation around this.

I think this is a good place to start.  Thanks for the work @chrisndodge.

@marcotuts: Would you have someone that could help provide UX review of some of the above points/questions?

chrisndodge commented:
@robrap thanks for the follow up.

1) The reason why they were made optional was purely to keep default behavior the same. Also I followed an existing pattern that was already in the codebase for similar LTI Launch Parameters.

2) Thanks for letting us know when there's a final decision on this

3) Will do

robrap commented:
Thanks @chrisndodge.  In regards to #1 above, the reason the other LTI launch parameters needed to have associated settings was because they contain PII (Personally Identifiable Information) and trigger an alert warning to the user.

The major reasons I can see for optional params are:
1. PII, or
2. Performance implications.

I'm not sure if #2 is a concern here or not.  Would either of these params trigger additional db calls, or is this just exposing cached data?

Also, I could imagine people wanting additional optional parameters over time, so it would seem like the settings interface might want a list of options with checkboxes, or something of that nature.  It feels both odd to bundle these two params under one setting, and natural of course because it happens to match this first use case.  Anyway, I was wondering if we could avoid this discussion altogether if we wouldn't pay a penalty for simply expanding the set of default params returned.

chrisndodge commented:
@robrap thank you.  In terms of DB-call overheads, with respect to making these default launch parameters:

For the retrieval of the user's enrollment mode (i.e. audit vs. verified vs. ...), we are calling into the existing xBlock service API endpoint in the 'credit' djangoapp (/openedx/core/djangoapps/credit/services.py). By eye there probably will be about  a half dozen or so MySQL queries that come out of that.

For the retrieval of the student's verification status, it calls the existing check_verify_status_by_course() which I believe is called by the Dashboard. That does some lookups as well. Likewise, maybe about half-dozen.

If need be I can bring up a Django Debug Toolbar to get more exact numbers.

Thanks again

robrap commented:
@chrisndodge: I think we should probably make these parameters optional as you had it, although it does add complexity for authors.  fyi: @ormsbee 

@marcotuts will be leading the way on the legal requirement for treatment for "External Resources" and will get back to us.  He's also planning on reviewing the UX for the settings for these new LTI launch parameters.

chrisndodge commented:
Thanks @robrap, let me know if there more information you all need.

marcotuts commented:
@chrisndodge - Would you mind sending me the sandbox credentials to review to my edx email? I can review and reply with product / UX notes after that. Thanks!

chrisndodge commented:
Hello @marcotuts, just sent sandbox information via email. Let me know if you have any questions.

robrap commented:
@chrisndodge - I'll also need credentials at some point, although I'm going
to wait on some design issues getting resolved with @marcotuts.  Thanks.

On Fri, Jul 29, 2016 at 4:18 PM, chrisndodge <notifications@github.com>
wrote:

> Hello @marcotuts <https://github.com/marcotuts>, just sent sandbox
> information via email. Let me know if you have any questions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/xblock-lti-consumer/pull/20#issuecomment-236283143>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AN5rPelriD-nbXvH-uABaguuSTKkEE8fks5qamAygaJpZM4JMfPD>
> .
>



-- 

*Robert Raposa*

edX | Software Architect  | rraposa@edx.org

141 Portland Street, 9th floor

Cambridge, MA 02139
http://www.edx.org <http://www.edxonline.org/>

[image:
http://www.e-learn.nl/media/blogs/e-learn/edX_Logo_Col_RGB_FINAL.jpg?mtime=1336074566]

marcotuts commented:
Hi Chris,

I've summarized a few thoughts below on requested changes based on the conversations we've had as a team. I will defer to @catong here on text string suggestions however. I've only included draft changes here when it comes to UI strings. 


**Fields and Settings**

1.  [Text Change] Request fields - 
![image](https://cloud.githubusercontent.com/assets/2023680/17308480/3aa250d0-5808-11e6-9433-15b1981ed565.png)

Setting Name: Request user's verification status 
Setting Help text: Select true to request a students course mode as well as their identity verification status.

2. [Text Change] Display Header
Setting name: Display Header
Setting help text: Select true to hide the header and grading information at the top of this component. 

3.) [Text Change] [(Optional) UI Change] Enable Custom HTML Launch Area
Two potential changes here. The smallest is to update the help text for this field to the following:

Setting help text: This setting allows for optional custom HTML to render as the launch pane, including the launch button. This only renders if the LTI Launch Target setting is set to 'Modal' or 'New Window.' 

Depending on level of effort, I would have suggested a setting called "Enable Custom HTML Launch Area" which would have exposed an "Editor" tab in the top of this component to mimic the pattern of other editor areas. I'm ok falling back to just the text change if this suggestion is a large change. Thanks !

![image](https://cloud.githubusercontent.com/assets/2023680/17308637/1861bd34-5809-11e6-8b55-6c1ff2e35d8e.png)

3.) External Resource Labeling
As an outcome of the discussion we've had with our legal / ux team here, this is where we landed. I've included images of where we are today for context.

Previously: 
![image](https://cloud.githubusercontent.com/assets/2023680/17308839/f489acc2-5809-11e6-9dd8-70e91840cafb.png)

With Changes: (image shows hidden header area option enabled)
![image](https://cloud.githubusercontent.com/assets/2023680/17308855/022af264-580a-11e6-9c85-4c18f40fe92f.png)

Final Outcome / Change (image shows header, but it could be hidden entirely as in the image above):
![image](https://cloud.githubusercontent.com/assets/2023680/17308900/2204c13c-580a-11e6-870b-b4fa3bc5e555.png)

- The gray box shows the area for the default HTML launch area which would include the buttons / text. To be clear, I'm not suggesting this box needs to be gray, it just is a stand-in for that area of the component. 
- In the bottom right of this area is where we would add the tertiary support text label to denote the external resource nature of this block. This echoes the pattern we've set up for conveying similar legal details and information, such as the license information we show on videos / blocks / pages. 
(Example license reference at the bottom of a page:) 
![image](https://cloud.githubusercontent.com/assets/2023680/17309044/d23acbe6-580a-11e6-98e0-59c1b5e79732.png)
(Example of a license detail on a video:)
![image](https://cloud.githubusercontent.com/assets/2023680/17309061/ee6bfe84-580a-11e6-9861-6f4a778ac9b8.png)

chrisndodge commented:
Thanks @marcotuts for the feedback. I'll huddle with the rest of the team here, and I'll let you know when we have revisions up for review.

robrap commented:
@marcotuts: The other item we discussed, but did not conclude, is whether it makes sense to unbundle the two new LTI launch parameters with one setting per parameter.  

I understand from an ASU authoring perspective the interest in a UI that makes their process as simple as possible.  However, I could also imagine a more generic list of parameters/checkboxes to include that would be simpler to maintain and grow without having to consider various bundles that may not be wanted or required by everyone.

@chrisndodge.  Thoughts?

chrisndodge commented:
@marcotuts thanks for the suggested text changes. To avoid multiple passes at text edits, I'll wait for feedback from @catong as well.

@marcotuts @robrap in terms of edits to the "Settings Panel" - there are two suggestions, one to change the list of parameters to include in the launch parameters so they aren't broken out one by one, the other to put the HTML edit pane on a separate tab - I'll review those with the ASU team.

catong commented:
@chrisndodge Thanks! We are meeting with @marcotuts tomorrow to get some background, then @lamagnifica will be most likely be taking this PR as she is more familiar with LTI than I am.

lamagnifica commented:
Hello hello @chrisndodge! Here are a few tweaks to Marco's suggestions above.

Question: how does the current "Button Text" field work with the new "Enable Custom HTML Launch Area" field? Is any supplied button text ignored? (that is my working assumption).  The help text below the Button Text field might also need to be updated... I took a stab at it below.

Question: It appears that there is some way to group fields other than in alphabetical order (!) is that true? I will note where I think each of the new fields should appear (but I can devise labels that put them in the right place alphabetically if need be...)

1.) Transmit user's course_mode...

- Setting Name: Request user's enrollment data 
- Setting Help text: Select True to request enrollment track and identity verification status information for users.
- Placement: after the other two "Request user's..." fields is fine

2.) Display Header Setting

- Setting name: Show Display Name
- Setting help text: Select True to show both the learner's grade and the Display Name for this component in the LMS. 
- Placement: after the Display Name field

3.) Enable Custom HTML Launch Area

- Setting name: Provide HTML to Customize Launch Pane
- Setting help text: Applies only if LTI Launch Target is set to Modal or New Window. Select True to enter HTML for the launch information, including the launch button, that appears to users. 
- Placement: after the LTI Launch Target field

4.) Launch Pane HTML

- Setting name: Launch Pane HTML Editor
- Setting help text: Enter HTML for the launch information, including the launch button, that appears to users. 
- Placement: after the Provide HTML to Customize Launch Pane field

External Resource Labeling
- For the text in the bottom right of Marco's proposed area, update the string to "This button opens an external resource." (if a screen reader will always identify the linked control as a button, otherwise use  "link" instead of "button")

Button Label: (just in case this is useful)
Setting help text: Enter the text on the button used to launch the third party application. Applies only if Hide External Tool is set to False, LTI Launch Target is set to Modal or New Window, and  custom launch pane HTML does not include a launch button. 

Please don't hesitate to tag me with any questions etc.

chrisndodge commented:
@lamagnifica thanks for the finalized text suggestions

@robrap in order to consolidate iterations of edits, do you have any python related comments? Thanks.

robrap commented:
@chrisndodge: I'll review the code that seems more stable.

Some outstanding questions:
1. You mentioned you'd get back about the idea of splitting out the LTI launch params.  It feels like the momentum of the original solution is simply taking over.
2. You mentioned that you'd get back with the steps required for making the custom HTML actually work.  Did I miss them in an earlier comment?  At a minimum, I imagine this will affect help text or documentation.  Also, as I mentioned, it may make sense to validate as well.

Thanks.

robrap commented on lti_consumer/lti.py:
```diff
@@ -159,6 +159,32 @@ def get_signed_lti_parameters(self):
 
+            credit_service = self.xblock.runtime.service(self.xblock, 'credit')
+            user_id = getattr(real_user_object, "id", None)
+            if self.xblock.transmit_course_mode_and_status and credit_service and user_id:
```
Would it make sense to add some else conditions as appropriate with warning or info logs for anything unexpected to help track down issues?

chrisndodge replied:
@robrap one slightly unfortunate thing here is that when the xBlock is loaded in Studio it does not have the full set of runtime support as it does when it is loaded into LMS. For example, we don't have access to a user_id, as I recall. I'll add a few logging items, but just because some of these settings might not be present in the runtime, it doesn't necessarily  make it an exception. But - as you say - more debugging information could be helpful for diagnostics.

robrap replied:
Got it. I'd say add what ever you think would be more helpful than unhelpful. Thanks.

robrap commented on lti_consumer/static/js/xblock_lti_consumer.js:
```diff
@@ -7,7 +10,7 @@ function LtiConsumerXBlock(runtime, element) {
             iframeModal: function (options) {
                 var $trigger = $(this);
-                var modal_id = $trigger.data("target");
+                var modal_id = "#"+$element.find(".lti-modal").attr("id");
```
Add spaces around `+`.

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -1,18 +1,23 @@
+    % if has_score and weight:
+        <div class="problem-progress">
+            % if module_score is not None:
+                (${"{points} / {total_points} points".format(points=module_score, total_points=weight)})
```
Not sure to what extent you want to fix lack of i18n and/or safe template fixes on code you are touching.

chrisndodge replied:
@robrap I guess I could be a "good citizen" and address a few prior existing i18n/safe template issues. If something gets overly burdensome, I'll bring it up.

robrap replied:
Do what makes sense.  There were a couple of places with new strings, like the aria-label, so at least cover new code.  Any other small fixes that move us in the right direction would be appreciated.  Thanks again.

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -1,18 +1,23 @@
+% if display_header:
+    <h2 class="problem-header">
+        ## Translators:  "External resource" means that this learning module is hosted on a platform external to the edX LMS
+        ${display_name} (External resource)
```
I'm getting confused by this code.  Does translation ever happen if no call to _() is made?  Shouldn't this be something like?:
```
${_('{display_name} (External resource)').format(display_name=display_name)}
```
Assuming this is correct, there is missing translation code throughout.  Is something different done with xblocks?

robrap replied:
Also, either the entire template should be made safe by default, or you should temporarily add HTML escaping filters like `${_('...') | h}` on translations.  

See more details here: http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#making-legacy-mako-templates-safe-by-default

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -22,25 +27,31 @@ <h2 class="problem-header">
-                    </button>
+        <section
+            class="wrapper-lti-link"
+            aria-label="This section of the courseware is delivered by a 3rd party. As such, we cannot make any guarantees that this external resource supports the same level of accessibility."
```
i18n?

robrap replied:
I didn't mark every string. Maybe you can review for this before the next pass?

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -22,25 +27,31 @@ <h2 class="problem-header">
+                        <button
+                            class="btn btn-pl-primary btn-base btn-lti-modal"
+                        >
+                            ${button_text or 'View resource in a modal window'} <i class="icon fa fa-external-link"></i>
```
i18n

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -22,25 +27,31 @@ <h2 class="problem-header">
+                            class="btn btn-pl-primary btn-base btn-lti-new-window"
+                            data-target="${form_url}"
+                        >
+                            ${button_text or 'View resource in a new window'} <i class="icon fa fa-external-link"></i>
```
i18n

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -22,25 +27,31 @@ <h2 class="problem-header">
+            aria-label="This section of the courseware is delivered by a 3rd party. As such, we cannot make any guarantees that this external resource supports the same level of accessibility."
+        >
+            % if launch_pane_html:
+                ${launch_pane_html}
```
Note, if you do try making the template safe by default, here you would need:
```
${launch_pane_html | n, decode.utf8}
```
It wouldn't hurt making this change now, which will make future work easier.

robrap commented on lti_consumer/tests/unit/test_lti.py:
```diff
@@ -176,16 +176,39 @@ def test_get_signed_lti_parameters(self):
         self.lti_consumer.xblock.has_score = True
         self.lti_consumer.xblock.ask_to_send_username = True
         self.lti_consumer.xblock.ask_to_send_email = True
+        self.lti_consumer.xblock.transmit_course_mode_and_status = True
```
What about other cases in the code like:
- transmit_course_mode_and_status = False
- services unavailable
- Other return values (e.g. get_course_verification_status => None)
It seems that at a minimum, covering the first case can be done in a test with all these params set to false and ensuring that none of the optional params appear.

robrap commented:
@chrisndodge: Generally, the heart of the code looks clean and simple.  Thank you.

As mentioned above, do what makes sense on cleanup of legacy code.  Let's just not add any new issues.  Thanks again.

cdodge-am commented:
@robrap, thanks for the review. I should be able to address the feedback pretty quickly. Thanks again.

edx-webhook commented:
Thanks for the pull request, @chrisndodge! I've created [OSPR-1380](https://openedx.atlassian.net/browse/OSPR-1380) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)     or indicated your institutional affiliation.     If you like, you can add yourself to the [AUTHORS](https://github.com/edx/xblock-lti-consumer/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

chrisndodge commented:
@lamagnifica I just noticed that I didn't answer one of your questions up above:

**"Question: how does the current "Button Text" field work with the new "Enable Custom HTML Launch Area" field? Is any supplied button text ignored? (that is my working assumption). The help text below the Button Text field might also need to be updated... I took a stab at it below."**

If the Custom HTML Launch Area is used, then the Button Text is ignored. Does that change any of your text recommendations?

lamagnifica commented:
@chrisndodge thanks, that was a helpful clarification. The very last item in my long comment above should work then.

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -36,7 +37,9 @@ <h2 class="problem-header">
-                ${launch_pane_html}
+                ## it is intended that authors can use HTML in the launch_pane_html parameter
+                ## so we allow this to go through unescaped
+                ${launch_pane_html | n}
```
Please update to:
```
${launch_pane_html | n, decode.utf8}
```
See http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#mako-filter-ordering-and-the-n-filter

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -22,25 +32,35 @@ <h2 class="problem-header">
+            % if launch_pane_html:
+                ## it is intended that authors can use HTML in the launch_pane_html parameter
+                ## so we allow this to go through unescaped
+                ${launch_pane_html | n}
```
Please update to:
```
${launch_pane_html | n, decode.utf8}
```
See http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#mako-filter-ordering-and-the-n-filter

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -22,25 +32,35 @@ <h2 class="problem-header">
+                % if description:
+                    ## I believe historically, we have some courses which pipe through some
+                    ## HTML in the description field, so let's preserve existing behavior
+                    <div class="lti-description">${description | n}</div>
```
@chrisndodge:
```
${description | n, decode.utf8}
```
@chrisndodge: I think the above is fine for now.  Below is just a discussion for ongoing efforts, not necessarily for this PR.

FYI: @nedbat @e0d @nasthagiri: Here is another place where people were using HTML because we didn't have any limits, so it has become a feature.  
How do we determine:
1. Whether or not to intentionally leave these features and possibly update docs to explicitly mention that HTML is allowed?
2. Do we want some sort of bleaching in these cases where it isn't a choice between no HTML and any HTML, but some limited set of safe(er) HTML?

chrisndodge replied:
@robrap if you can point me to an bleaching example, I'll see if I can add it now. I presume the main concern is with JS getting through, right?

robrap replied:
@nedbat @e0d @nasthagiri: I'm going to answer my earlier question.  I think we should be more explicit if we wish to allow HTML, and we should just use bleach with defaults as we have in other places.
@lamagnifica: We may want to mention that a limited set of HTML is acceptable for the description, but I'll leave that up to you.

chrisndodge commented:
@robrap thanks for the ongoing feedback, but I have a few edits today, so it might be easier if I just ping with you an "all set" message when I think I've hit  your feedback points. I also have to update my sandbox. Thanks again.

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -1,18 +1,28 @@
+<%!
+from django.utils.translation import ugettext as _
+%>
+<%page expression_filter="h"/>
```
@chrisndodge: I appreciate you going for the safe template work on this.
At the bottom of this file, you'll find:
```
   ## sanitized with bleach in view
   ${comment}
```
I assume that this also takes HTML and will need the `| n, decode.utf8` filter.  Also, since this mentions sanitization and bleach, maybe you could add the same comment/view sanitization code to description to keep it safe as well?

robrap commented:
@chrisndodge: Sorry if I missed the response, but I asked earlier:
> Could you provide steps/requirements for filling in the "Launch Pane HTML" such that it will work? It would be nice for there to be some sort of validation around this.

Is the answer simply that the HTML must include a button with the class "btn-lti-modal"?  Other?  Again, it would probably be nice to have onscreen docs for this (sorry if I missed it) and to have validation.
FYI: @lamagnifica @marcotuts

robrap commented:
@chrisndodge: FYI: I'm on vacation from Wed 17th (in 1.5 days) through the following Wed.  I'm not sure if this can sit that long.  If not, I may need to find someone else to help with any final review needs.  Please let me know.  Thanks!

robrap commented:
@chrisndodge: I'll hold up on additional comments until you give the word.  Just note my note about vacation.

Here is the sanitization line:
https://github.com/chrisndodge/xblock-lti-consumer/blob/49c3e82c1ef2570bfa7f5327ceec1aab745e8968/lti_consumer/lti_consumer.py#L856

nedbat commented:
@chrisndodge Let us know if there's something we can do on our end to resolve the legal status of this contribution.  I know we are going to get it straightened out, but we should wait to merge this pull request until it is.

chrisndodge commented:
@robrap thanks for the heads up regarding your upcoming vacation. I will try to wrap up this round of edits by start of day tomorrow.

chrisndodge commented:
@nedbat I've checked in with ASU about the status of the Entity Contributor Agreement. We should be able to get back to you shortly - it might have been sent already.

cdodge-am commented:
@robrap regarding your question on making the launch button, yes it is just a matter of using a specific class name. I'll update the documentation.

chrisndodge commented:
@robrap regarding your question on making the launch button, yes it is just a matter of using a specific class name. I'll update the documentation.

aricken commented:
I got an update about the ECA. Was sent in Friday. Edx had a couple
minor questions. ASU is addressing. New draft should be back to edX within
1-2 days.

On Monday, August 15, 2016, chrisndodge <notifications@github.com> wrote:

> @nedbat <https://github.com/nedbat> I've checked in with ASU about the
> status of the Entity Contributor Agreement. We should be able to get back
> to you shortly - it might have been sent already.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/xblock-lti-consumer/pull/20#issuecomment-239985730>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFJzarRaw5tgFq7Gis67omPg3egdqSWQks5qgSF2gaJpZM4JMfPD>
> .
>

chrisndodge commented:
@robrap I think we need a final decision on "bleaching" of the description field. I'm almost certain that ASU currently has HTML in that field with a live course, so adding bleaching will probably adversely affect it.

AFAIK, the plan is to switch over from putting HTML in the description field and migrate it over to use this new launch_pane_html field. However there will be a short overlap in production before the courseware content can be migrated.

So this might have to be a two phase migration.

chrisndodge commented:
@robrap also I wanted to call attention to the fact there is a companion PR in edx-platform to extend one of the existing xBlock services: PR#12997

Both PRs need to be merged for this submission to be considered complete.

Thanks

chrisndodge commented:
@robrap OK, I think I've addressed all feedback from everyone, but the comments have been spread out a bit.

I need to update the sandbox with latest code, but I wanted to send a signal that I think I'm done with code related changes (unless I missed some comments). I'll send followup note when sandbox has been refreshed.

chrisndodge commented:
@robrap @lamagnifica @marcotuts , sandbox has been refresh with latest code

robrap commented:
@chrisndodge: @dianakhuang will be taking over with backup from @douglashall starting tomorrow.  There may not be anything left by then.  If it is ready to merge and the agreement is in, you can decide together whether to merge before my return end of next week.

robrap commented on lti_consumer/tests/unit/test_lti.py:
```diff
@@ -190,7 +217,101 @@ def test_get_signed_lti_parameters(self):
+
+        self.assertEqual(self.lti_consumer.get_signed_lti_parameters(), expected_lti_parameters)
+
+        # Test that `lis_person_sourcedid`, `lis_person_contact_email_primary`, and `launch_presentation_locale`
```
Comment is missing the new params.  Maybe update comment to:
```
# Test that the following LTI parameters are not included when a user cannot be found.
```

robrap commented on lti_consumer/tests/unit/test_lti.py:
```diff
@@ -204,21 +325,59 @@ def _mock_services(_, service_name):
-        self.lti_consumer.xblock.runtime.service = _mock_services
+        self.lti_consumer.xblock.runtime.service = _mock_verification_none
+
+        print self.lti_consumer.get_signed_lti_parameters()
```
Code you meant to delete?

robrap commented on lti_consumer/tests/unit/test_lti.py:
```diff
@@ -190,7 +217,101 @@ def test_get_signed_lti_parameters(self):
+        PropertyMock(return_value=('t', 's'))
+    )
+    @patch('lti_consumer.lti_consumer.LtiConsumerXBlock.user_id', PropertyMock(return_value=FAKE_USER_ID))
+    def test_get_signed_lti_parameters_none_verification_status(self):
```
Seems like a lot of repeated code both in the patch decorators and inside these test functions.  Could you remove some of the redundancy?  Maybe a custom decorator to package the patch decorators, and/or any other solution?

chrisndodge replied:
@robrap I spent a little time trying to figure out how to do this, but it exceeds my python-foo. Do you have a suggestion?

robrap replied:
@calen was the one who mentioned this was possible, but I wouldn't know how without researching.  Maybe @cpennington can answer?  I'm going to be out of the country later today.  Good luck!

cpennington replied:
@chrisndodge: You have a couple of options:

1. `@patch` the class (which will apply the same decorators to every test method)
2. Write your own decorator that bundles all of the existing decorators:

    def decorators_for_tests(func):
        func = decorator1(args)(func)
        func = decorator2(args)(func)
        ...
        return func

robrap commented:
@chrisdodge: Thanks for the additions.  I've completed my pass.  The following comments are related to earlier comments:

1. In terms of the bleaching question, is there a way to find out if it really causes an issue?  I'd be curious what HTML they are using that bleaching would hurt.  Is there any way to get your hands on some sample HTML, even if it is just to note for a future story regarding completing the safe template work?

2. You still need to implement this comment regarding `${comment}.`: https://github.com/edx/xblock-lti-consumer/pull/20#discussion_r74778232

chrisndodge commented:
@robrap (although I know you are on vacation)

#1: I don't have staff access to the live ASU Algebra course, but in the actual page in the LMS, I believe the HTML that is in use for the "description" field is:

<center><font size="5" color="#4f4f4f"><b>Embrace challenge + master <font size="5" color="#00acc8">College Algebra</font></b></font><br></center>

#2: Addressed that comment

chrisndodge commented:
@robrap I think you are back from vacation, so I wanted to touch base on this PR. Are there any open issues here? Also anything remaining on https://github.com/edx/edx-platform/pull/12997?

If we're all set here, let go over merge/deploy timing. Thanks.

robrap commented:
@chrisndodge: I'm back.  Could you send me credentials for the sandbox?  Thanks!
@marcotuts: Are you all set for merge?
@lamagnifica: Are you all set for merge?  Note some text updates after your last review.

robrap commented:
@chrisndodge: Sorry about late comments on the UX, but this is first I looked at the sandbox.  I'll let you and @marcotuts decide if any of these comments warrant fixing/changing:

- Original studio editor fields are ordered in alphabetical order.  This is no longer the case, and thus the order varies a lot from before this PR.  I'd recommend bringing back the old order to minimize changes, but I'll let you both decide.  Also, if not alphabetical, you may want to still adjust the current order.
- The title `Launch Pane HTML Editor` should probably read `Custom Launch Pane HTML`.
- Assuming that the following help text is accurate, I would prepend the help text with the following sentence: `Leave blank for the default launch pane. For a custom launch pane, enter HTML for...`. 

FYI: @lamagnifica 

Also, I still plan on doing some local testing before merge.

cdodge-am commented:
@robrap I believe @lamagnifica made the suggestion to re-order the fields. I can certainly change it back.

I'll wait for confirmation from @marcotuts about those other two suggestions.

Thanks

robrap commented:
Thanks @chrisndodge.  I had missed that.  No need to revert the order.  
@lamagnifica @marcotuts: Can you verify you are happy with the order of the fields as they appear in the sandbox?  Thanks.

lamagnifica commented:
apologies, I was heads down with Ned all afternoon working on Eucalyptus. I will review first thing Monday if that is all right.

robrap commented on lti_consumer/templates/html/student.html:
```diff
@@ -22,28 +32,41 @@ <h2 class="problem-header">
+                    ## HTML in the description field, so let's preserve existing behavior
+                    <div class="lti-description">${description | n, decode.utf8}</div>
+                % endif
+                <p class="lti-link external">
```
What do the classes `lti-link` and `external` do?  Do you need them with the custom html?

The reason I was looking at this is because my manual testing found the following defect:
The custom pane HTML with new window (versus modal) seems to be broken.

Also, what will happen if the class on the button doesn't properly match the "LTI Launch Target" setting?  Do we need a textual reminder to have them match?  It would be better if the author didn't have to be concerned with all this, and if something could be done (maybe in JavaScript) to simply add the appropriate class based on the LTI Launch Target.

robrap commented:
@lamagnifica @marcotuts: Reminder that there are some comments above that need your review.  You may also want to review the interaction between the "LTI Launch Target" option and the class that the author needs to add the the custom Launch Pane HTML ([see comment)](https://github.com/edx/xblock-lti-consumer/pull/20#discussion_r76666794).

@chrisndodge: You can see the [broken custom launch here](http://studio-asu.chrisndodge.com/container/block-v1:LTI+LTI+LTI+type@vertical+block@f16229b85e9e429cb540586c23c3c4fd).

lamagnifica commented:
Hi @chrisndodge thanks for your patience, and thanks @robrap for the ping.
I'd like to suggest the following updates. I don't think any of them are show stoppers.

- For the text under the "Launch Pane HTML Editor" field, I suggest streamlining the second sentence from 
In order to make the launch button functional, be sure to add a class='btn-lti-modal' or class='btn-lti-new-window' to the button element. 
to
The button element must include a class='btn-lti-modal' or class='btn-lti-new-window'. 

- Please change 
Button Text
to
Button Label

- I also noticed the text in the template yml file (?) that appears when you first add the component. If possible, change 
Please provide launch_url. Click "Edit", and fill in the required fields. 
to
To provide the URL of the external tool and other required information, select Edit.

chrisndodge commented:
@robrap thanks for the heads up. Will look into it right away.

@lamagnifica thanks for the feedback.

robrap commented:
@lamagnifica: I had made some language related suggestions in this comment: https://github.com/edx/xblock-lti-consumer/pull/20#issuecomment-242841858
I'm not sure if you missed them and agree or disagree, or if you saw them and disagreed, but didn't state your disagreement explicitly.
Thanks.

lamagnifica commented:
@robrap, I did see your comment, but did not understand that you were waiting for my input.

> @chrisndodge: Sorry about late comments on the UX, but this is first I looked at the sandbox. I'll let you and @marcotuts decide if any of these comments warrant fixing/changing:
> 
>     Original studio editor fields are ordered in alphabetical order. This is no longer the case, and thus the order varies a lot from before this PR. I'd recommend bringing back the old order to minimize changes, but I'll let you both decide. Also, if not alphabetical, you may want to still adjust the current order.
(AH: I think the alpha order issue for this was discussed and resolved. Correct? Do you have a different field order to propose? I'd be happy to give feedback. Or did you want me to propose a different field order?)
>     The title Launch Pane HTML Editor should probably read Custom Launch Pane HTML.
(AH: I partially disagree, since I think the word "Editor" is important for this label. If you think "Custom" is important, ok by me to add it, though this is getting pretty lengthy.)
>     Assuming that the following help text is accurate, I would prepend the help text with the following sentence: Leave blank for the default launch pane. For a custom launch pane, enter HTML for....
(AH: I didn't understand what you are indicating by "the following help text" before, but I spent some more time in the sandbox and I *think* you mean, add more explicit info about usage to the help text under the "custom launch pane HTML editor" field. If so, I think it's a good idea. See below.)
> 
> FYI: @lamagnifica
> 
> Also, I still plan on doing some local testing before merge.
> 


Looking at the Studio sandbox to understand Rob's last question made me realize that I probably misunderstood whether a launch button with a class is a required part of the custom launch pane html. It now seems to me that it is optional to include a button for the custom launch pane, and that if you don't include a button, that's when the Button Text (aka Button Label) field comes in to play. But, if you DO include a button, you need to specify its class. So.

"To override the default launch pane that appears to users, add the HTML for your customized launch information. If you include a launch button, the button element must include either class='btn-lti-modal' or class='btn-lti-new-window'. "

robrap commented:
@lamagnifica: regarding order, I had missed your earlier comment when I wrote mine, and @chrisndodge pointed it out, so we are all set with the order you proposed.

The Launch Pane HTML is optional.  If you leave it blank, you get the default button.  If you fill it in, it MUST contain a button which MUST contain the classname that matches the choice made in "LTI Launch Target".  This field is optional, because there is a default.  I was trying to allude to that with the word "Custom".  It's fine if you want to drop "Custom" and you think the help text is enough to show it is optional.  Another possibility is replacing "Pane" with "Button", as in: "Launch Button HTML Editor".

At this point, I'll leave it to whatever you recommend, @lamagnifica.  I just wanted to clear up how it works so you can be clear about the necessary title and help text.  Thanks.

marcotuts commented:
@chrisndodge:
I'm reviewing this again and one thing I missed is where / how the "Enable Custom HTML Editing" setting disappeared and whether it is now the case that all LTI blocks need to define their HTML Launch Area. Is that an accurate statement on my part? This approach leads in part to @robrap's concern around needing to enter in certain classes to author this correctly. 

The concern here is that we've now made basic launching of an LTI block more complex in the service of flexible HTML authoring for all cases. 

I'm reviewing this tomorrow morning with @robrap to make sure I'm not missing any context already from the reviews that are now collapsed away. 

Thanks!

lamagnifica commented:
@robrap and @marcotuts loop me back in after you decide what fields are included.

cdodge-am commented:
@marcotuts @robrap @lamagnifica thanks for feedback. Right now I'm on a quasi-vacation week. I'll most likely get back to you right after labor day when I'm back at my development machine. Thx.

marcotuts commented:
@chrisndodge -  I have had a chance to catch up with the team to understand the history of changes here since I last reviewed thi. There is one area that I believe still needs to be addressed. 

 **Class Dependency**
We should eliminate the need to cross-update html / configuration as is currently required via the btn-lti-modal or the btn-lti-new-window classes. Simply requiring that authors include a button should be enough for this block to appropriately add classes (or use whatever other mechanism necessary) to ensure the proper behavior based on other configuration. What are your thoughts on level of complexity on this? Should that change be pushed to a second follow-up PR? 

---
There are cases that would also benefit from validation here, as this would let authors know if for example, they haven't included a button element in their html (only relevant when custom HTML area is in use.) We can punt on validation for now. 

---
Separately, I believe the LTI Consumer tool should be reviewed / updated after a UX audit / review, as I'm concerned about the authoring complexity and misconfiguration potential. No timeline on this effort, and it is outside the scope of this pull request. 

Thanks Chris! (and all other reviewers as well!)

chrisndodge commented:
Hello @marcotuts @robrap @lamagnifica. I'm back at my development machine and getting ready to address the latest round of questions.

Just to clarify, are the remaining things to look at are:

- How to auto-inject the required class name on any button

- Validation on the customized HTML

- Investigate the lack of correct behavior of customized HTML with a designed launch behavior or being a new window (as reported by @robrap)

Is that correct? Thanks.

marcotuts commented:
@chrisndodge  - Thanks for the summary here. That seems correct as a list, though I think we are ok to split off validation to another story since that would open up a set of questions around workflow, UI strings, etc. I've included thoughts below on validation anyways, but it could be considered a follow-on PR.  

Re: Validation (#2 in your list) - It would be great to emulate the validation we use on the Library Content (aka Randomized Content Block) where we show you a message in case of a configuration error. Ex: "There is no launch button included in your custom HTML area."  Is this the only validation case to consider? cc'ed @robrap

shaunagm commented:
Hi @chrisndodge, what's the status on this PR?

chrisndodge commented:
Hi @shaunagm thank you for the ping and your attentiveness. I'm going to review remaining recommendations/pieces of feedback. I'll post a follow-up comment early next week.

shaunagm commented:
@chrisndodge, did you get a chance to review this?  No rush on our end, just checking in.

gsong commented:
@chrisndodge @marcotuts What do we want to do with this PR? Keep moving it forward or close it out?

edx-webhook commented:
Hello @chrisndodge: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.

gsong commented:
Closing due to inactivity.

robrap commented:
FYI: @marcotuts
